console/collections/src/sparse_kary_merkle_tree/mod.rs

```diff
@@ -0,0 +1,290 @@
+
+        // Start with the leaf hash.
+        let mut current_hash = *self.leaves.get(key_hash).ok_or_else(|| anyhow!("Leaf not found"))?;
+
+        // Store the leaf at its path position in the nodes map as well.
+        // This allows other paths to find this leaf when they need it as a sibling.
+        let leaf_path = path_indices.iter().map(|&i| u8::try_from(i)).collect::<Result<Vec<_>, _>>()?;
+        self.nodes.insert(leaf_path, current_hash);
+
+        // Traverse from the leaf to the root, updating all nodes along the path.
+        for depth in (0..DEPTH).rev() {
+            let index = path_indices[depth as usize];
+
+            // Compute the sibling hashes at this level.
+            let mut children = Vec::with_capacity(ARITY as usize);
+            for i in 0..ARITY {
+                if i as usize == index {
+                    children.push(current_hash);
+                } else {
+                    // Try to get the sibling from the nodes map, otherwise use empty hash.
+                    let sibling_path = Self::compute_node_path(&path_indices[..depth as usize], i)?;
+                    children.push(*self.nodes.get(&sibling_path).unwrap_or(&self.empty_hash));
+                }
+            }
+
+            // Hash the children to get the parent hash.
+            let parent_hash = self.path_hasher.hash_children(&children)?;
+
+            // Store the PARENT hash at its path (the path from root to this parent node).
+            // The parent is at depth `depth`, represented by path_indices[..depth].
+            if depth > 0 {
+                let parent_path =
+                    path_indices[..depth as usize].iter().map(|&i| u8::try_from(i)).collect::<Result<Vec<_>, _>>()?;
+                self.nodes.insert(parent_path, parent_hash);
+            }
+
+            current_hash = parent_hash;
+        }
+
+        // Update the root.
+        self.root = current_hash;
+
+        Ok(())
+    }
+
+    /// Computes the path indices through the tree for a given key hash.
+    /// Uses consecutive bits to maximize entropy. With sufficient depth, this provides
+    /// strong collision resistance (e.g., DEPTH=32, ARITY=2 uses 32 bits → 2^32 positions).
+    fn compute_path_indices(&self, key_hash: &Field<E>) -> Result<Vec<usize>> {
+        let mut indices = Vec::with_capacity(DEPTH as usize);
+
+        // Convert the key hash to bits.
+        let key_bits = key_hash.to_bits_le();
+
+        // Compute the number of bits needed per level.
+        let bits_per_level = ARITY.next_power_of_two().trailing_zeros() as usize;
+
+        // Use CONSECUTIVE bits from the key hash to get maximum entropy.
+        // For example: DEPTH=32, ARITY=2 → uses bits [0..32] → 2^32 unique positions.
+        for depth in 0..DEPTH as usize {
+            let start_bit = depth * bits_per_level;
+            let end_bit = std::cmp::min(start_bit + bits_per_level, key_bits.len());
```

Comment:
Should we `ensure!` that the number of bits we get from `key_hash` doesn't exceed the required `DEPTH*bits_per_level`? Otherwise `end_bit` might end up lower than `start_bit`.